drake/systems/plants/RigidBodyManipulator.h

```diff
@@ -97,46 +99,256 @@ class DLLEXPORT_RBM RigidBodyManipulator
+                auto dSdotVdvi = dSdotv.rightCols(joint.getNumVelocities());
+                djoint_acceldv.rightCols(nv_joint) += transformSpatialMotion(element.transform_to_world, dSdotVdvi);
+
+                auto dJdotVdv = element.motion_subspace_in_world_dot_times_v.gradient().value().rightCols(num_velocities);
+                auto dJdotVdv_parent = parent_element.motion_subspace_in_world_dot_times_v.gradient().value().rightCols(num_velocities);
+                dJdotVdv.setZero();
+                for (int j = 0; j < nv_branch; j++) {
+                  int v_index = v_indices[j];
+                  dJdotVdv.col(v_index) = dJdotVdv_parent.col(v_index) + djoint_acceldv.col(j);
+                }
+              }
+            }
+          }
+        }
+      }
+      else {
+        element.transform_to_world.setIdentity();
+        // motion subspace in body frame is empty
+        // motion subspace in world frame is empty
+        // qdot to v is empty
+        // v to qdot is empty
+        if (compute_gradients) {
+          // gradient of transform
+          element.dtransform_to_world_dq.setZero();
+          // gradient of motion subspace in world is empty
+        }
+        if (cache.hasV()) {
+          element.twist_in_world.value().setZero();
+          element.motion_subspace_in_body.value().setZero();
+          element.motion_subspace_in_world.value().setZero();
+          element.qdot_to_v.value().setZero();
+          element.v_to_qdot.value().setZero();
+
+          if (compute_gradients) {
+            element.twist_in_world.gradient().value().setZero();
+            element.motion_subspace_in_body.gradient().value().setZero();
+            element.motion_subspace_in_world.gradient().value().setZero();
+            element.qdot_to_v.gradient().value().setZero();
+            element.v_to_qdot.gradient().value().setZero();
+          }
+          if (compute_JdotV) {
+            element.motion_subspace_in_body_dot_times_v.value().setZero();
+            element.motion_subspace_in_world_dot_times_v.value().setZero();
+            if (compute_gradients) {
+              element.motion_subspace_in_body_dot_times_v.gradient().value().setZero();
+              element.motion_subspace_in_world_dot_times_v.gradient().value().setZero();
+            }
+          }
+        }
+      }
+    }
+
+    cache.setJdotVCached(compute_JdotV && cache.hasV());
+  };
+
+  bool isBodyPartOfRobot(const RigidBody& body, const std::set<int>& robotnum) const;
+
+  double getMass(const std::set<int>& robotnum = RigidBody::defaultRobotNumSet) const;
 
   template <typename Scalar>
-  GradientVar<Scalar, SPACE_DIMENSION, 1> centerOfMass(int gradient_order, const std::set<int>& robotnum = RigidBody::defaultRobotNumSet);
+  GradientVar<Scalar, SPACE_DIMENSION, 1> centerOfMass(KinematicsCache<Scalar>& cache, int gradient_order, const std::set<int>& robotnum = RigidBody::defaultRobotNumSet) const;
```

Comment:
Nitpick:  The way that you are using this -- as a now mandatory argument for all of the kinematics functions, i don't really like the name KinematicsCache.  I think KinematicsSolution and kinsol would be more appropriate.  

Another organization that we could consider is having the KinematicsCache/Sol keep a pointer to the RBM, and then to actually move most of these methods over to the KinematicsCache class.  That would, I think, really be a cleaner interface.  
```
kinsol = doKinematics(q)
kinsol->forwardKin()
kinsol->centerOfMass()
...
```
That way you don't have to carry around both the kinsol and the rbm and pass them in together every time.  And it would safeguard against the possibility of someone passing a kinsol for the wrong RBM into these methods.